### DISCUSSION TITLE
Multi-camera/view of same data

### DISCUSSION THREAD

**Discussion by ASDFGerte** [pixijs/pixijs#8567] [Q&A]
I am moderately new to pixi, and am trying to figure out the meta for having multiple camera-views of the same data. While having found something that sounds feasible, as I am rather uncertain, and would appreciate a quick "yeah that's not too absurd", to prevent running down the wrong path.

When trying to render two views of the same scene (e.g. split-screen), my initial idea would have been to render two different sub-trees of the same data (nodes being any container-like object, leaf-nodes sprites):
![SceneGraph](https://user-images.githubusercontent.com/23001266/183663563-8a441d8c-2aad-4c2d-9617-4e48e956a822.png)

However, the overlap is apparently not easy to do (or impossible?). The scene-graph in pixijs in my understanding is a scene-tree, and each element may only have one parent. Following, I considered three ways, however each being more or less obviously inefficient:
 - Clone the data for every view. This becomes unfeasible, if there are really many sprites, or the existing method of creating them is expensive (in my case, there are up to ~100k sprites on screen, per view, and creating each requires a bit of logic to determine the texture. Note that's an estimated upper bound, and some optimizations may be added later, e.g. caching groups with `RenderTexture`s).
 - Keep one structure, which contains everything, and move/render that for every view. This means not culling objects, which are known to not be rendered for the respective view.
 - On every frame, detach subtrees, attach them to the other view, render, detach and re-attach to the original one. This is both doing a lot of useless operations, as well as being difficult to keep track of.

However, for the second point, I can also do culling by not removing the objects from the container, but setting `renderable`/`visible` accordingly. This is the current best I could find so far, and without any new insights, likely what will end up getting used.

Is that how it should be done? Do you have any further remarks about the topic?

---

A related question I could find is https://github.com/pixijs/pixijs/discussions/8356, which however has no answers.

**Comment by domis86** [comment]
What about https://github.com/davidfig/pixi-viewport/issues/182#issuecomment-565769167 ?

**Reply by ASDFGerte** [reply]
That there is a library for viewport is interesting, thanks. It is indeed related. However, for the suggestions:
 - The first is to copy everything, which I already described. Perhaps it is premature optimization, but having tens of thousands of sprites in memory N times, although they don't differ, sounds really awkward.
 - The second is similar to my current conclusion, one render-pass per view. I am unsure how that library handles culling, perhaps I'll investigate later.

It also gives me more indication, that there is nothing too obvious, which I'd have missed.
